inox2d/src/render.rs

```diff
@@ -218,3 +222,173 @@ impl Puppet {
+                self.set_mask_mode(mask.mode == MaskMode::Dodge)?;
+
+                let mask_node = puppet.nodes.get_node(mask.source).unwrap();
+                let mask_node_render_ctx = &puppet.render_ctx.node_render_ctxs[&mask.source];
+
+                match (&mask_node.data, &mask_node_render_ctx.kind) {
+                    (
+                        InoxData::Part(ref mask_part),
+                        RenderCtxKind::Part(ref mask_part_render_ctx),
+                    ) => {
+                        self.draw_part_self(
+                            true,
+                            camera,
+                            mask_node_render_ctx,
+                            mask_part,
+                            mask_part_render_ctx,
+                        )?;
+                    }
+
+                    (
+                        InoxData::Composite(ref mask_composite),
+                        RenderCtxKind::Composite(ref mask_children),
+                    ) => {
+                        self.draw_composite(true, camera, mask_composite, puppet, mask_children)?;
+                    }
+
+                    _ => {
+                        // This match block clearly is sign that the data structure needs rework
+                        todo!();
+                    }
+                }
+            }
+            self.on_begin_masked_content()?;
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+            self.on_end_mask()?;
+        } else {
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+        }
+        Ok(())
+    }
+
+    fn begin_composite(&self) -> Result<(), Self::Error>;
+    fn end_composite(&self) -> Result<(), Self::Error>;
+
+    fn draw_composite_self(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error> {
+        self.begin_composite()?;
+
+        for &uuid in children {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            if let (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) =
+                (&node.data, &node_render_ctx.kind)
+            {
+                if as_mask {
+                    self.draw_part_self(true, camera, node_render_ctx, part, part_render_ctx)?;
+                } else {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+            } else {
+                // composite inside composite simply cannot happen
+            }
+        }
+
+        self.end_composite()?;
+        Ok(())
+    }
+
+    fn draw_composite(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        composite: &Composite,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
```

Comment:
These are the two considerations behind the designs of these 4 render functions:

0. `draw_part` and `draw_composite` are the top level draw functions, which when called (with `as_mask` set to `false`) on a fresh canvas, should yield one complete drawable.
1. `draw_part_self` and `draw_composite_self` are respectively, components that get called by `draw_part` `draw_composite`. They "belong" to the larger outside function, providing one step of the drawing process. In specific, they are named with `_self` because they draw the very content of the drawable in question"
    - For `Part`, `draw_part_self` draws the mesh warped texture of a `Part`.
    - For `Composite`, `draw_composite_self` draws all children content.

    However the outer part exist because only drawing the very content belongs to self does not complete the drawing. For `Part`, `Mask`s need to be deployed before drawing, and for `Composite`, composited content needs to be transferred to a correct location.

Then I can explain why certain functions are default, or to say I want to provide users with a ready-to-use fixed solution:

- For `Part`, as mask related functions are implemented elsewhere, and how `Mask` interact with `Part` is defined in the Inochi2D standard, we can just call them in a predefined order to deploy masks. What we don't know without renderer-specific knowledge is how to actually render a texture warped by mesh. We can almost say this step is the basic building block of all visual results. So `draw_part` is defined by us, calling `draw_part_self` and mask related functions, with `draw_part_self` implemented by the user.
- For `Composite`, the main content is then compositing all the content of its children, who are all `Part`s. However as all `Part` rendering is defined elsewhere, we already know how to draw all children. What we don't know without renderer-specific knowledge is how to actually transfer the composited content to the correct location. So `draw_composite_self` is defined by us, being called by `draw_composite`, with `draw_composite` implemented by the user.

Summarizing, we can give these 4 functions completely different names so that no one will be confused about one `_self` is default while the other one is not. Anyway the above is almost the explanation for how Inochi2D renders.

Reply:
Hmmm, I see.

So about the suffixes, how about `draw_part_mesh` and `draw_composite_children`? It's slightly more descriptive (even though at the end of the day all these functions are gonna be documented).

Reply:
I have gone out of this mind game and figured out that the trick is not to try to give `Part` and `Composite` related methods similar names. Users will have to implement `draw_part_content` because we don't know how the rendering should be happening. Users also will have to implement `prepare_composite_draw` and `transfer_composite_content` (these are the only two composite-related methods the user have to implement) because we don't know how the composite buffer is prepared and read after drawing is complete. Now the names are more descriptive and users won't mess them up.
And we do our job by calling the three methods in the right way. Users won't see how we name the internal functions anyway.

Reply:
I see. I think I like this solution better. Though as a last suggestion, what about `begin_composite_content` as opposed to `prepare_composite_draw`? The pair of methods you have to implement would be the same and still distinct from `draw_part_content` so I think it's gonna be fine.